From: olh@suse.de
Subject: correct keymapping on Powerbook built-in USB ISO keyboards

similar to the version in adbhid_input_register()
Corrects also the castrated Powerbooks.

---
 drivers/usb/input/hid-core.c  |    3 +++
 drivers/usb/input/hid-input.c |   10 +++++++++-
 drivers/usb/input/hid.h       |    1 +
 3 files changed, 13 insertions(+), 1 deletion(-)

Index: linux-2.6.18/drivers/usb/input/hid-core.c
===================================================================
--- linux-2.6.18.orig/drivers/usb/input/hid-core.c
+++ linux-2.6.18/drivers/usb/input/hid-core.c
@@ -1742,6 +1742,9 @@ static const struct hid_blacklist {
 	{ USB_VENDOR_ID_APPLE, 0x030A, HID_QUIRK_POWERBOOK_HAS_FN },
 	{ USB_VENDOR_ID_APPLE, 0x030B, HID_QUIRK_POWERBOOK_HAS_FN },
 
+	{ USB_VENDOR_ID_APPLE, 0x0215, HID_QUIRK_POWERBOOK_ISO_KEYBOARD },
+	{ USB_VENDOR_ID_APPLE, 0x0218, HID_QUIRK_POWERBOOK_ISO_KEYBOARD },
+
 	{ USB_VENDOR_ID_PANJIT, 0x0001, HID_QUIRK_IGNORE },
 	{ USB_VENDOR_ID_PANJIT, 0x0002, HID_QUIRK_IGNORE },
 	{ USB_VENDOR_ID_PANJIT, 0x0003, HID_QUIRK_IGNORE },
Index: linux-2.6.18/drivers/usb/input/hid-input.c
===================================================================
--- linux-2.6.18.orig/drivers/usb/input/hid-input.c
+++ linux-2.6.18/drivers/usb/input/hid-input.c
@@ -232,6 +232,7 @@ static void hidinput_configure_usage(str
 	struct hid_device *device = input->private;
 	int max = 0, code;
 	unsigned long *bit = NULL;
+	unsigned char swap_key;
 
 	field->hidinput = hidinput;
 
@@ -255,7 +256,14 @@ static void hidinput_configure_usage(str
 
 			if ((usage->hid & HID_USAGE) < 256) {
 				if (!hid_keyboard[usage->hid & HID_USAGE]) goto ignore;
-				map_key_clear(hid_keyboard[usage->hid & HID_USAGE]);
+				swap_key = usage->hid & HID_USAGE;
+				if (device->quirks & HID_QUIRK_POWERBOOK_ISO_KEYBOARD) {
+					if (swap_key == 100)
+						swap_key = 53;
+					else if (swap_key == 53)
+						swap_key = 100;
+				}
+				map_key_clear(hid_keyboard[swap_key]);
 			} else
 				map_key(KEY_UNKNOWN);
 
Index: linux-2.6.18/drivers/usb/input/hid.h
===================================================================
--- linux-2.6.18.orig/drivers/usb/input/hid.h
+++ linux-2.6.18/drivers/usb/input/hid.h
@@ -260,6 +260,7 @@ struct hid_item {
 #define HID_QUIRK_POWERBOOK_HAS_FN		0x00001000
 #define HID_QUIRK_POWERBOOK_FN_ON		0x00002000
 #define HID_QUIRK_INVERT_HWHEEL			0x00004000
+#define HID_QUIRK_POWERBOOK_ISO_KEYBOARD	0x00008000
 
 /*
  * This is the global environment of the parser. This information is
